pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+			// requeued implicitly when we update our status with the new error
+			// condition. If not, we want to try again after a short wait.
+			managed.SetConditions(v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+		}
+
+		if err := r.managed.Complete(ctx, managed); err != nil {
+			// If this is the first time we encounter this issue we'll be
+			// requeued implicitly when we update our status with the new error
+			// condition. If not, we want to try again after a short wait.
+			managed.SetConditions(v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+		}
+
+		// We've successfully created our external resource. In many cases the
+		// creation process takes a little time to finish. We requeue a short
+		// wait in order to observe the external resource to determine whether
+		// it's ready for use.
+		managed.SetConditions(v1alpha1.ReconcileSuccess())
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	u, err := external.Update(ctx, managed)
+	if err != nil {
+		// We'll hit this condition if we can't update our external resource,
+		// for example if our provider credentials don't have access to update
+		// it. If this is the first time we encounter this issue we'll be
+		// requeued implicitly when we update our status with the new error
+		// condition. If not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	if err := r.managed.Publish(ctx, managed, u.ConnectionDetails); err != nil {
+		// If this is the first time we encounter this issue we'll be requeued
+		// implicitly when we update our status with the new error condition. If
+		// not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	// We've successfully attempted to update our external resource. This will
+	// often be a no-op. Per the below issue nothing will notify us if and when
+	// the external resource we manage changes, so we requeue a speculative
+	// reconcile after a long wait in order to observe it and react accordingly.
+	// https://github.com/crossplaneio/crossplane/issues/289
+	managed.SetConditions(v1alpha1.ReconcileSuccess())
+	return reconcile.Result{RequeueAfter: r.longWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+}
+
+// An APISecretPublisher publishes ConnectionDetails by creating a Secret.
+type APISecretPublisher struct {
```

Comment:
In this context `API` is supposed to communicate that `APISecretPublisher` publishes secrets by acting as a client of a Kubernetes API server. It's effectively the 'default' implementation of the `ManagedPublisher` interface, but I prefer to avoid names like `DefaultSecretPublisher` and thus tried to find something succinct that (kind of) explains what variant it is.